pkg/agent/apiserver/handlers/ovstracing/handler.go

```diff
@@ -0,0 +1,299 @@
+		traceReq.SrcIP = ip
+		if inPort == nil {
+			// Input port not specified. Try using the source OVS port.
+			inPort = intf
+		}
+	}
+
+	gatewayConfig := aq.GetNodeConfig().GatewayConfig
+	if req.destination != nil {
+		ip, intf, err := getPeerAddress(aq, req.destination)
+		if err != nil && err.StatusCode == http.StatusNotFound && req.destination.name != "" {
+			// The destination might be a Service.
+			ip, err = getServiceClusterIP(aq, req.destination.name, req.destination.namespace)
+		}
+		if err != nil {
+			return nil, err
+		}
+		if ip == nil {
+			return nil, handlers.NewHandlerError(errors.New("destination has no IP address"), http.StatusNotFound)
+		}
+		traceReq.DstIP = ip
+		if intf != nil {
+			// Must be a local Pod or OVS port. Use interface MAC as the packet
+			// destination MAC.
+			traceReq.DstMAC = intf.MAC
+		} else {
+			// Should be a remote Pod or IP. Use gateway MAC as the destination MAC.
+			traceReq.DstMAC = gatewayConfig.MAC
+		}
+	}
+
+	if inPort == nil {
+		if req.source != nil && req.source.name != "" {
+			// Source is a remote Pod. Use the default tunnel port as the input port.
+			// For hybrid TrafficEncapMode, even the remote Node is in the same subnet
+			// as the source Node, the tunnel port is still used as the input port.
+			intf, ok := aq.GetInterfaceStore().GetInterface(config.DefaultTunPortName)
+			// If the default tunnel port is not found, it might be NoEncap or
+			// NetworkPolicyOnly mode. Use gateway port as the input port then.
+			if ok {
+				inPort = intf
+			}
+		}
+		// Use gateway port as the input port when the source is an IP address
+		// (assuming it is an external IP or a Node IP).
+	}
+	if inPort != nil {
+		if inPort.Type == interfacestore.ContainerInterface {
+			traceReq.SrcMAC = inPort.MAC
+		} else if inPort.Type == interfacestore.TunnelInterface {
+			// Use tunnel traffic virtual MAC for both source and destination MAC
+			// addresses of the trace packet input from the tunnel port.
+			traceReq.SrcMAC = aq.GetOpenflowClient().GetTunnelVirtualMAC()
+			traceReq.DstMAC = traceReq.SrcMAC
+		} else if inPort.InterfaceName == gatewayConfig.Name {
+			traceReq.SrcMAC = gatewayConfig.MAC
+		} else {
+			return nil, handlers.NewHandlerError(errors.New("invalid OVS port"), http.StatusBadRequest)
+		}
+		traceReq.InPort = inPort.InterfaceName
+	} else {
+		// Use gateway port as the input port if it could not be figured out from the
+		// source.
+		traceReq.InPort = gatewayConfig.Name
+		traceReq.SrcMAC = gatewayConfig.MAC
+	}
+
+	return &traceReq, nil
+}
+
+// parseTracingPeer parses Pod/Service name and Namespace or OVS port name or
+// IPv4 address from the string. nil is returned if the string is not of a
+// valid Pod/Service reference ("Namespace/name") or OVS port name format, and
+// not an IPv4 address.
+func parseTracingPeer(str string) *tracingPeer {
+	parts := strings.Split(str, "/")
+	n := len(parts)
+	if n > 2 {
```

Comment:
Same comment as for a previous PR. I do believe that at some point in the future, we should use different API parameters for the different cases, and determine which parameter to use client-side in antctl.

Reply:
Yes. But if you think ok, I would not do the antctl framework change in this PR.